components/script/dom/gpucanvascontext.rs

```diff
@@ -131,17 +190,93 @@ impl GPUCanvasContext {
     }
 }
 
+// Abstract ops from spec
+impl GPUCanvasContext {
+    /// <https://gpuweb.github.io/gpuweb/#abstract-opdef-gputexturedescriptor-for-the-canvas-and-configuration>
+    fn texture_descriptor_for_canvas(
+        &self,
+        configuration: &GPUCanvasConfiguration,
+    ) -> GPUTextureDescriptor {
+        let size = self.size();
+        GPUTextureDescriptor {
+            format: configuration.format,
+            // We need to add `COPY_SRC` so we can copy texture to presentation buffer
+            // causes FAIL on webgpu:web_platform,canvas,configure:usage:*
+            usage: configuration.usage | GPUTextureUsageConstants::COPY_SRC,
+            size: GPUExtent3D::GPUExtent3DDict(GPUExtent3DDict {
+                width: size.width as u32,
+                height: size.height as u32,
+                depthOrArrayLayers: 1,
+            }),
+            viewFormats: configuration.viewFormats.clone(),
+            // other members to default
+            mipLevelCount: 1,
+            sampleCount: 1,
+            parent: GPUObjectDescriptorBase {
+                label: USVString::default(),
+            },
+            dimension: GPUTextureDimension::_2d,
+        }
+    }
+
+    /// <https://gpuweb.github.io/gpuweb/#abstract-opdef-expire-the-current-texture>
+    fn expire_current_texture(&self) {
+        if let Some(current_texture) = self.current_texture.take() {
+            // Make copy of texture content
+            self.send_swap_chain_present(current_texture.id());
+            // Step 1
+            current_texture.Destroy()
+        }
+    }
+
+    /// <https://gpuweb.github.io/gpuweb/#abstract-opdef-replace-the-drawing-buffer>
+    fn replace_drawing_buffer(&self) {
+        // Step 1
```

Comment:
As a side note, I increasingly find step numbers less useful than just copying the actual spec sentence(see for example what we're currently doing with the readablestream part: https://github.com/servo/servo/blob/8542f7253cdaec3b88914d7e206c19055f87b09f/components/script/dom/readablestreamdefaultcontroller.rs#L285)